From: Philipp Zabel <p.zabel@pengutronix.de>
Date: Mon, 16 Mar 2015 15:19:29 +0100
Subject: [PATCH] [media] coda: add cache bypass module parameters

Allow to disable the cache for encoding and decoding via the
disable_encoder_cache and disable_decoder_cache module parameters.

Signed-off-by: Philipp Zabel <p.zabel@pengutronix.de>
---
 drivers/media/platform/coda/coda-bit.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/drivers/media/platform/coda/coda-bit.c b/drivers/media/platform/coda/coda-bit.c
index 4763276b58c9..35b9418afb49 100644
--- a/drivers/media/platform/coda/coda-bit.c
+++ b/drivers/media/platform/coda/coda-bit.c
@@ -16,6 +16,7 @@
 #include <linux/irqreturn.h>
 #include <linux/kernel.h>
 #include <linux/log2.h>
+#include <linux/module.h>
 #include <linux/platform_device.h>
 #include <linux/reset.h>
 #include <linux/slab.h>
@@ -40,6 +41,14 @@
 #define CODA_DEFAULT_GAMMA	4096
 #define CODA9_DEFAULT_GAMMA	24576	/* 0.75 * 32768 */
 
+static int disable_encoder_cache = 0;
+module_param(disable_encoder_cache, int, 0644);
+MODULE_PARM_DESC(disable_encoder_cache, "Disable 2D cache for encoder");
+
+static int disable_decoder_cache = 0;
+module_param(disable_decoder_cache, int, 0644);
+MODULE_PARM_DESC(disable_decoder_cache, "Disable 2D cache for decoder");
+
 int tiled_output = 0;
 module_param(tiled_output, int, 0644);
 MODULE_PARM_DESC(tiled_output, "Output tiled buffers from the decoder even though the rotator doesn't work");
@@ -788,6 +797,7 @@ err_clk_per:
 static void coda9_set_frame_cache(struct coda_ctx *ctx, u32 fourcc)
 {
 	u32 cache_size, cache_config;
+	bool disable_cache;
 
 	if (ctx->tiled_map_type == GDI_LINEAR_FRAME_MAP) {
 		/* Luma 2x0 page, 2x6 cache, chroma 2x0 page, 2x4 cache size */
@@ -808,6 +818,12 @@ static void coda9_set_frame_cache(struct coda_ctx *ctx, u32 fourcc)
 				8 << CODA9_CACHE_CR_BUFFER_SIZE_OFFSET |
 				8 << CODA9_CACHE_CB_BUFFER_SIZE_OFFSET;
 	}
+	if (ctx->inst_type == CODA_INST_ENCODER)
+		disable_cache = disable_encoder_cache;
+	else
+		disable_cache = disable_decoder_cache;
+	if (disable_cache)
+		cache_config |= BIT(CODA9_CACHE_BYPASS_OFFSET);
 	coda_write(ctx->dev, cache_config, CODA9_CMD_SET_FRAME_CACHE_CONFIG);
 }
 
